src/main/java/com/kimho/book/service/impl/BookServiceImpl.java

```diff
@@ -0,0 +1,124 @@
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import java.util.Date;
+import java.util.List;
+
+@Service
+public class BookServiceImpl implements BooksService<BookDto, BookUpdate> {
+    @Autowired
+    private BookRepository bookRepository;
+
+    @Autowired
+    private BookToDto bookToDto;
+
+    @Autowired
+    private DtoToBook dtoToBook;
+
+    @Autowired
+    private UserServiceImpl userService;
+
+    @Override
+    public List<BookDto> getAll() {
+        List<Book> books = bookRepository.findAll();
+        return bookToDto.convert(books);
+    }
+
+    @Override
+    public BookDto get(long id) {
+        verifyBookIsExist(id);
+        return bookToDto.convert(bookRepository.findById(id).get());
+    }
+
+    private Book findById(long id) {
+        verifyBookIsExist(id);
+        return bookRepository.findById(id).get();
+    }
+
+    @Override
+    public BookDto add(BookDto bookDto) {
+        User myUser = userService.getMyUser();
+        bookDto.setUserId(userService.getMyUser().getId());
+        Book book = dtoToBook.convert(bookDto);
+        if (myUser.isAdmin() || myUser.isSuperAdmin()) {
+            book.setEnabled(true);
+        }
+        BookDto bookResponse = bookToDto.convert(bookRepository.save(book));
+        return bookResponse;
+    }
+
+    @Override
+    public BookDto edit(BookUpdate bookEdition, long id) {
+        verifyBookIsExist(id);
+        Book book = bookRepository.findById(id).get();
+        if (!checkAuthorizationEditBook(book)) {
+            throw new UnauthorizedException("Unauthorized");
+        }
+        book.setTitle(bookEdition.getTitle());
+        book.setAuthor(bookEdition.getAuthor());
+        book.setDescription(bookEdition.getDescription());
+        book.setImage(bookEdition.getImage());
+        book.setUpdatedAt(new Date());
+        return bookToDto.convert(bookRepository.save(book));
+    }
+
+    @Override
+    public void delete(long id) {
+        verifyBookIsExist(id);
+        if (!checkAuthorizationEditBook(findById(id))) {
+            throw new UnauthorizedException("Unauthorized");
+        }
+        bookRepository.deleteById(id);
+    }
+
+    public BookDto enable(long id) {
+        Book book = bookRepository.getById(id);
+        if (!checkAuthorizationEnableBook(book)) {
+            throw new UnauthorizedException("Unauthorized");
+        }
+        book.setEnabled(!book.isEnabled());
+        return bookToDto.convert(bookRepository.save(book));
+    }
+
+    private boolean checkAuthorizationEditBook(Book book) {
+        User myUser = userService.getMyUser();
+        User userCreateBook = book.getUser();
+        if (userCreateBook.getId() == myUser.getId()) {
+            return true;
+        } else if (myUser.isSuperAdmin()) {
+            return true;
+        } else if (userCreateBook.isUser() && myUser.isAdmin()) {
+            return true;
+        }
```

Comment:
nếu là admin và super admin thì có thể edit book chứ e?
`userCreateBook.isUser() && myUser.isAdmin()` : điều kiện nj là sao e?

Reply:
Nếu là admin thì chỉ được quyền sửa xóa bài viết của user 
Không được sửa xóa bài viết của admin khác

Reply:
admin chỉ được change user và chính nó thôi anh
admin không change được admin khác

Reply:
admin có quyền edit book của user khác chứ e?
admin toàn quyền e nhé